### PR TITLE
Tutorial kool

### PR DESCRIPTION
**Pull request by daejunpark**

### CODE REVIEW

**Comment by daejunpark** [comment]
_[NOTE: This PR includes #1754. Once it is merged, I'll rebase this PR, so that the changes will be much cleaner.]_

I added todo comments in the following form:

```
// TODO(KORE): <description>
<current ugly rules>
/*
<original rules>
*/
```

Issues are:
- major
  - parsing issues
  - cell fragment issues
- minor
  - unsupported latex annotations
  - implicit subsort issue
  - sort inferences
  - context bug
  - getKLabel issue

I'll open issues for each of above.

**Comment by daejunpark** [comment]
@bmmoore @grosu please review

**Comment by bmmoore** [comment]
Well, that's pretty ugly. Good work showing us where problems are.

The commented-out rules at lines 945 and 997 don't seem to match any nearby TODO comment.

Why did you remove the concrete syntax production for `frange` rather than just declaring and using a KLabel but leaving the original production, like you did for `farray` and `fcall`?

Does every kind of TODO have an open issue? It would be good to put issue numbers in the definition (maybe in a glossary at the top instead of adding to each TODO?).

Several that don't remind me of open issues are the comment `TODO: remove the sort cast ::Id of C above, when sort inference bug fixed` at line 1054, and the `TODO(KORE): clarify sort inferences` where you moved productions into `Exp` from `KItem`, and the "parsing issues" (I also suspect that fcall conflicts with label applications, but I would expect `farray` and `frange` to work).
that there are `parsing issues` that interfere with the `frange` or `farray` productions. (The

File: k-distribution/tutorial/2_languages/2_kool/1_untyped/kool-untyped.k [hunk 1]
```diff
@@ -347,15 +364,21 @@ object and method closures. */
...
 definition. */
 
+  // TODO(KORE): clarify sort inferences
+  syntax Exp ::= "undefined"  [latex(\bot)]
+  /*
   syntax KItem ::= "undefined"  [latex(\bot)]
+  */
 
-  rule <k> var X:Id; => . ...</k>
+  rule <k> var X:Id; => .K ...</k>
```
Comment (grosu) [inline k-distribution/tutorial/2_languages/2_kool/1_untyped/kool-untyped.k:382]: TODO: Change .K back to .

File: k-distribution/tutorial/2_languages/2_kool/1_untyped/kool-untyped.k [hunk 2]
```diff
@@ -408,14 +443,18 @@ definition. */
 
   rule V1:Val == V2:Val => V1 ==K V2
   rule V1:Val != V2:Val => V1 =/=K V2
-  rule ! T => notBool(T)
+  rule ! T:Bool => notBool(T)
```
Comment (grosu) [inline k-distribution/tutorial/2_languages/2_kool/1_untyped/kool-untyped.k:451]: TODO: Casting should not be needed, fix it once the sort inferencer works.

File: k-distribution/tutorial/2_languages/2_kool/1_untyped/kool-untyped.k [hunk 3]
```diff
@@ -445,17 +484,17 @@ interestingly, the semantics of return stays unchanged. */
 
   context (HOLE => lvalue(HOLE)) = _
 
-  rule <k> loc(L) = V:Val => V ...</k> <store>... L |-> (_ => V) ...</store>
+  rule <k> loc(L) = V:Val => V ...</k> <store>... L |-> (_:Exp => V) ...</store>
```
Comment (grosu) [inline k-distribution/tutorial/2_languages/2_kool/1_untyped/kool-untyped.k:496]: TODO: casting of _ should not be needed

File: k-distribution/tutorial/2_languages/2_kool/1_untyped/kool-untyped.k [hunk 4]
```diff
@@ -494,43 +537,47 @@ semantics of thread spawning will be defined shortly.  However,
 interestingly, the other concurrency constructs keep their semantics
 from SIMPLE unchanged. */
 
-  rule (<thread>... <k>.</k> <holds>H</holds> <id>T</id> ...</thread> => .)
+  rule `<thread>... <k>.K</k> <holds>H</holds> <id>T</id> ...</thread> => .Bag`
```
Comment (grosu) [inline k-distribution/tutorial/2_languages/2_kool/1_untyped/kool-untyped.k:553]: TODO: ..Bag should be .

**Comment by daejunpark** [comment]
Added issue numbers in TODOs. The outstanding issues are:
- #1682 List Parsing adding terminator badly
- #1705 Need subsorting between related lists.
- #1744 Cell Fragments Related Issue
- #1747 cell fragment parsing issue
- #1748 cell fragment as an argument
- #1772 On .K, .Map, .List, .KList ...
- #1799 Support latex annotations.
- #1801 partial functions in symbolic rewriting
- #1802 context kompilation bug in kore
- #1803 sort inference in kore

@bmmoore regarding `frange`, it turned out to be a little bit tricky, so that I want to change back to original once #1682 is fixed.

**Comment by daejunpark** [comment]
@bmmoore please review again.

**Comment by daejunpark** [comment]
@bmmoore I've cut the dependency with #1754, and rebased on the current master. Now it is safe to merge. Please review and merge if it looks good.

**Comment by bmmoore** [comment]
Looks good. I made one more comment about a TODO, but as long as the definition actually compiles and runs I agree we should merge it to master and fix the todos there.

**Pull request merged**